### ISSUE TITLE
Whisper Tokenizer Experiments

### ISSUE THREAD

**Issue by gangagyatso4364** [OpenPecha/garchen-rinpoche-speech-sig#31]
<h2>🧩 Whisper Tokenizer Experiments</h2>
<h3>🧠 Description</h3>
<p>The <strong>Whisper model</strong> uses a <strong>byte-level BPE (Byte Pair Encoding)</strong> tokenizer similar to GPT-2’s, trained on large multilingual datasets.<br>
This tokenizer starts from UTF-8 bytes (0–255) and merges frequent byte sequences into subword tokens. It includes:</p>
<ul>
<li>
<p><strong>Byte-level BPE tokens</strong> — learned multilingual subwords</p>
</li>
<li>
<p><strong>Language tokens</strong> (e.g. <code inline="">&lt;|en|&gt;</code>, <code inline="">&lt;|bo|&gt;</code>, <code inline="">&lt;|zh|&gt;</code>)</p>
</li>
<li>
<p><strong>Task tokens</strong> (<code inline="">&lt;|transcribe|&gt;</code>, <code inline="">&lt;|translate|&gt;</code>)</p>
</li>
<li>
<p><strong>Timestamp and control tokens</strong> (<code inline="">&lt;|notimestamps|&gt;</code>, <code inline="">&lt;|startoftranscript|&gt;</code>, <code inline="">&lt;|endoftext|&gt;</code>)</p>
</li>
</ul>
<p>While this architecture allows Whisper to process any language, its <strong>Tibetan coverage is limited</strong> because the original training corpus contained <strong>very little Tibetan text</strong>.<br>
As a result:</p>
<ul>
<li>
<p>Tibetan words are split into many byte-level fragments instead of meaningful syllables</p>
</li>
<li>
<p>The <strong>token sequences are longer</strong></p>
</li>
<li>
<p>The model <strong>requires more compute and time</strong> during inference</p>
</li>
<li>
<p>ASR accuracy (WER/CER) drops compared to high-resource languages</p>
</li>
</ul>
<p>This experiment systematically tests Whisper’s <strong>ASR performance</strong> using different <strong>tokenizer setups</strong> and <strong>transcript formats</strong> to identify the most efficient configuration for Tibetan.<br>
Each setup varies in tokenizer type (default, extended, custom BPE, etc.) and transcript language (Tibetan or Wylie), allowing direct comparison of <strong>recognition accuracy</strong>, <strong>token efficiency</strong>, and <strong>inference speed</strong>.</p>
<hr>
<h3>🎯 Completion Criteria</h3>
<p>A comprehensive comparison of <strong>Whisper ASR performance</strong> across multiple tokenizer and transcript setups, with detailed evaluation of:</p>
<ul>
<li>
<p><strong>Word Error Rate (WER)</strong></p>
</li>
<li>
<p><strong>Character Error Rate (CER)</strong></p>
</li>
<li>
<p><strong>Average tokens per sentence</strong></p>
</li>
<li>
<p><strong>Inference speed and computational efficiency</strong></p>
</li>
</ul>
<hr>

<h3>🧪 Experiment Folders</h3>

<html><head></head><body>
Here are the tokenizer experiments:

| # | Model / Config | Tokenizer Type | Transcript | Folder Name |
| :-: | :-- | :-- | :-- | :-- |
| 1 | **Whisper Model** | Default Tokenizer | Wylie | `whisper_model_default_tokenizer_on_wylie` |
| 2 | **Whisper Model** | Default Tokenizer | Tibetan | `whisper_model_default_tokenizer_on_tibetan` |
| 3 | **Whisper Model** | Default + Added Tibetan Tokens | Tibetan | `whisper_model_added_tokens_tokenizer_on_tibetan` |
| 4 | **Whisper Model** | Custom Tibetan Tokens | Tibetan | `whisper_model_custom_tibetan_tokenizer_on_tibetan` |
| 5 | **Whisper Model** | Custom Byte-Level (Latin+BPE) Tokens | Tibetan | `whisper_model_custom_bpe_tokenizer_on_tibetan` |
| 6 | **Whisper Config (Random Init)** | Custom Byte-Level (Latin+BPE) Tokens | Tibetan | `whisper_config_custom_bpe_tokenizer_on_tibetan` |

</body></html>

<hr>
<h3>⚙️ Motivation Behind the Experiments</h3>
<ul>
<li>
<p>The <strong>default Whisper tokenizer</strong> is <strong>byte-level</strong> and <strong>language-agnostic</strong>, but its <strong>limited Tibetan subword merges</strong> make it inefficient.</p>
</li>
<li>
<p>This results in <strong>higher token counts</strong>, <strong>slower inference</strong>, and <strong>less accurate transcriptions</strong>.</p>
</li>
<li>
<p>By training <strong>Tibetan-aware tokenizers</strong> (custom BPE or syllable-level), we expect:</p>
<ul>
<li>
<p>Shorter token sequences</p>
</li>
<li>
<p>Faster inference and reduced compute</p>
</li>
<li>
<p>Better WER/CER alignment with ground truth</p>
</li>
<li>
<p>More linguistically meaningful tokenization</p>
</li>
</ul>

### ✅ Subtasks — Create Training Scripts

* [x] Create training script for `whisper_model_default_tokenizer_on_wylie`
* [x] Create training script for `whisper_model_default_tokenizer_on_tibetan`
* [x] Create training script for `whisper_model_added_tokens_tokenizer_on_tibetan`
* [x] Create training script for `whisper_model_custom_tibetan_tokenizer_on_tibetan`
* [x] Create training script for `whisper_model_custom_bpe_tokenizer_on_tibetan`
* [x] Create training script for `whisper_config_custom_bpe_tokenizer_on_tibetan`

</li>
</ul>
<hr>
<h3>👥 Reviewer</h3>
<ul class="contains-task-list">
<li class="task-list-item">
<p><input type="checkbox" checked="" disabled=""> <strong>@TenzinGayche</strong></p>
</li>
</ul>

**Comment by gangagyatso4364** [comment]
To see the effect of better vocab compare it with submited transcript corpus

**Comment by gangagyatso4364** [comment]
- [x] training script for whisper generic model on wylie transcript
- [x] training script for whisper  generic model on tibetan transcript
- [x] training script for whisper  model with added tibetan tokens on tibetan transcript
- [x] training script for whisper model with customised bytelevel tokens on tibetan transcript
- [x] training script for whisper architecture config (random weights) with customised byte level tokens on tibetan transcript
- [x] training script for whisper model with customised tibetan tokens on tibetan transcripts.

**Comment by ta4tsering** [comment]
so @gangagyatso4364 wasn't able to do the updated subtasks that are for custom Tibetan token train in unicode Tibetan and then latin too. He will wrap up the card here by adding all the trained models in the hugging face and then add the documentation.
@gangagyatso4364 need to do some evaluations using the existing code on vast.ai
He will complete the documentation and remaining work by the end of Friday tomorrow.

**Comment by ta4tsering** [comment]
Now we have 4 different tokenizer that will be used for both unicode and wylie, from which two are not applicable on wylie, so that means that we can use to whisper training on 6 different tokenizer.
Two are already trained out of 6, 4 still needs to be trained and tested.
Pipeline and the scripts are all ready to use at go but only need to use it for those 4 remaining.

@gangagyatso4364 please point to the card that will contain the 4 remaining model training in here.
Update the title of this card as well reflecting the implementations of this card.
